protocol/peer/peer.go

```diff
@@ -0,0 +1,108 @@
+package peer
+
+import (
+	"github.com/kaspanet/kaspad/netadapter"
+	"github.com/kaspanet/kaspad/util/daghash"
+	"github.com/kaspanet/kaspad/util/subnetworkid"
+	"github.com/kaspanet/kaspad/wire"
+	"github.com/pkg/errors"
+	"sync"
+	"sync/atomic"
+)
+
+// Peer holds data about a peer.
+type Peer struct {
+	ready uint32
+
+	selectedTipHashMtx sync.RWMutex
+	selectedTipHash    *daghash.Hash
+
+	id                 uint32
+	userAgent          string
+	services           wire.ServiceFlag
+	advertisedProtoVer uint32 // protocol version advertised by remote
+	protocolVersion    uint32 // negotiated protocol version
+	disableRelayTx     bool
+	subnetworkID       *subnetworkid.SubnetworkID
+}
+
+// SelectedTipHash returns the selected tip of the peer.
+func (p *Peer) SelectedTipHash() (*daghash.Hash, error) {
+	if atomic.LoadUint32(&p.ready) == 0 {
+		return nil, errors.New("peer is not ready yet")
+	}
+	p.selectedTipHashMtx.RLock()
+	defer p.selectedTipHashMtx.RUnlock()
+	return p.selectedTipHash, nil
+}
+
+// SubnetworkID returns the subnetwork the peer is associated with.
+// It is nil in full nodes.
+func (p *Peer) SubnetworkID() (*subnetworkid.SubnetworkID, error) {
+	if atomic.LoadUint32(&p.ready) == 0 {
+		return nil, errors.New("peer is not ready yet")
+	}
+	return p.subnetworkID, nil
+}
+
+// SetSelectedTipHash sets the selected tip of the peer.
+func (p *Peer) SetSelectedTipHash(hash *daghash.Hash) error {
+	if atomic.LoadUint32(&p.ready) == 0 {
+		return errors.New("peer is not ready yet")
+	}
+	p.selectedTipHashMtx.Lock()
+	defer p.selectedTipHashMtx.Unlock()
+	p.selectedTipHash = hash
+	return nil
+}
+
+// MarkAsReady marks the peer as ready.
+func (p *Peer) MarkAsReady() error {
+	if atomic.AddUint32(&p.ready, 1) != 1 {
+		return errors.New("peer is already ready")
+	}
+	return nil
+}
+
+// UpdateFieldsFromMsgVersion updates the peer with the data from the version message.
+func (p *Peer) UpdateFieldsFromMsgVersion(msg *wire.MsgVersion, peerID uint32) {
+	// Negotiate the protocol version.
+	p.advertisedProtoVer = msg.ProtocolVersion
+	p.protocolVersion = minUint32(p.protocolVersion, p.advertisedProtoVer)
+	log.Debugf("Negotiated protocol version %d for peer %s",
+		p.protocolVersion, p)
+
+	// Set the peer's ID.
+	p.id = peerID
+
+	// Set the supported services for the peer to what the remote peer
+	// advertised.
+	p.services = msg.Services
+
+	// Set the remote peer's user agent.
+	p.userAgent = msg.UserAgent
+
+	p.disableRelayTx = msg.DisableRelayTx
+	p.selectedTipHash = msg.SelectedTipHash
+	p.subnetworkID = msg.SubnetworkID
+}
+
+func (p *Peer) String() string {
+	//TODO(libp2p)
+	panic("unimplemented")
```

Comment:
Why not implement this?

Reply:
Because we're doing weird peerID stuff

Reply:
Alright, please make sure we have a ticket to take care of this / Stas knows he needs to implement this as part of what he does.